### ISSUE TITLE
Gate docstrings and type hints

### ISSUE THREAD

**Issue by Cryoris** [Qiskit/qiskit-terra#3705]
### What is the expected enhancement?

The standard gates could have a better documentation. They're missing type information on the input and only a few explain what their action on the qubits actually is. This forces users to go look either in the code (which some might not do) or elsewhere (books/internet). Since the qubit ordering is different in Qiskit compared to many other places, I believe it would be very useful to have a more accurate documentation. Especially for users getting started with quantum computing; the gates are probably one of the first objects they get in contact with. 

In #3472, many gates already get an updated docstring that includes the matrix representation. 
However the types are still missing and the docstrings could be more elaborate in general. 

For example we could go from
```
def rz(self, phi, q):  
    """Apply Rz to q."""
```
to
```

def rz(self, 
       phi: Union[int, Qubit], 
       qubit: Union[int, Qubit]) -> Instruction:
    r"""
    Apply the RZ gate to the specified qubit.

    The RZ gate rotates the state of the qubit around the Z-axis by the angle phi,
    e.g. it maps $|0\rangle$ to $|0\rangle$ and $|1\rangle$ to $-|1\rangle$.

    **Matrix Definition**
    
    The matrix for this gate is given by:
    
    .. math::
        U_{\text{RX}}(\theta)
            = \exp\left(-i \frac{\theta}{2} \sigma_X \right)
            = \begin{bmatrix}
                \cos(\theta / 2) & -i \sin(\theta / 2) \\
                -i \sin(\theta / 2) &  \cos(\theta / 2)
            \end{bmatrix}

    Args:
        phi: the rotation angle
        qubit: the qubit to which the RZ gate is applied

    Returns:
        the instruction added to self
    """
```
(Note: the matrix representation bit is part of #3472)

**Comment by Cryoris** [comment]
This is partially addressed by #4017 but it would probably still be useful for IDE and linter purposes to have type hints and type documentation in the docstring.

**Comment by MarkoShiva** [comment]
It is beneficial for whole project if type hints are supported.
